builtin/common/serialize.lua

```diff
@@ -1,205 +1,217 @@
-	-- content in local_defs.
-	function dump_or_ref_val(x)
-		if seen[x] ~= 2 then
-			return dump_val(x)
+-- Serializes Lua nil, booleans, numbers, strings, tables and even functions
+-- Tables are referenced by reference, strings are referenced by value. Supports circular tables.
+local function serialize(value, write)
+    -- References are stored as tables of characters A-Z, which are digits of a base 26 number
+	local reference = {"A"}
+	local function increment_reference(place)
+		if not reference[place] then
+			reference[place] = "B"
+		elseif reference[place] == "Z" then
+            -- Carriage
+			reference[place] = "A"
+			return increment_reference(place + 1)
+		else
+			reference[place] = succ[reference[place]]
 		end
-		local var = dumped[x]
-		if var then  -- Already referenced
-			return var
+	end
+    -- [object] = reference string
+	local references = {}
+    -- Circular tables that must be filled using `table[key] = value` statements
+	local to_fill = {}
+	for object, count in pairs(count_objects(value)) do
+		local type_ = type(object)
+        -- Object must appear more than once. If it is a string, the reference has to be shorter than the string.
+		if count >= 2 and (type_ ~= "string" or #reference + 2 < #object) then
+			local ref = table_concat(reference)
+			write(ref)
+			write("=")
+            if type_ == "table" then
+                write("{}")
+            elseif type_ == "function" then
+                write(dump_func(object))
+            elseif type_ == "string" then
+                write(quote(object))
+            end
+			write(";")
+			references[object] = ref
+			if type_ == "table" then
+				to_fill[object] = ref
+			end
+			increment_reference(1)
 		end
-		-- First occurence, create and register reference
-		local val = dump_val(x)
-		local i = local_index
-		local_index = local_index + 1
-		var = "_["..i.."]"
-		local_defs[#local_defs + 1] = var.." = "..val
-		dumped[x] = var
-		return var
 	end
-
-	-- Second phase.  Dump the object; subparts occuring multiple times
-	-- are dumped in local variables which can be referenced multiple
-	-- times.  Care is taken to dump local vars in a sensible order.
-	function dump_val(x)
-		local  tp = type(x)
-		if     x  == nil        then return "nil"
-		elseif tp == "string"   then return string.format("%q", x)
-		elseif tp == "boolean"  then return x and "true" or "false"
-		elseif tp == "function" then
-			return string.format("loadstring(%q)", string.dump(x))
-		elseif tp == "number"   then
-			-- Serialize numbers reversibly with string.format
-			return string.format("%.17g", x)
-		elseif tp == "table" then
-			local vals = {}
-			local idx_dumped = {}
-			local np = nest_points[x]
-			for i, v in ipairs(x) do
-				if not np or not np[i] then
-					vals[#vals + 1] = dump_or_ref_val(v)
-				end
-				idx_dumped[i] = true
+	local function use_short_key(key)
+		return not references[key] and type(key) == "string" and string_match(key, "^[%a_][%a%d_]*$")
+	end
```

Comment:
Pls comment this func.
Ie. something like:
`-- Used to decide whether we can do "key=..." instead of "["..quote(key).."]=..."`

Reply:
```lua
					if use_short_key(k) then
						write(k)
					else
						write("[")
						dump(k)
						write("]")
					end
```

Is quite self-explaining, isn't it?

Reply:
Yes, but it comes much later in code. Many functions are self-explaining if you look up how they are used, but this looking up is work.

Reply:
Well, I guess a comment wouldn't hurt